documentation/api/query/v4/pql.markdown

```diff
@@ -0,0 +1,361 @@
+> backend, so the available
+> [regexp features](http://www.postgresql.org/docs/9.4/static/functions-matching.html#POSIX-SYNTAX-DETAILS)
+> are determined by PostgreSQL.  For best results, use the simplest
+> and most common features that can accomplish your task.
+
+#### Array Match: `in`
+
+The `in` operator matches a field, or set of fields against either an array or a subquery.
+
+The `in` operator can be used in two ways. The simplest way is to see if a field, contains one of the values
+provided in a list of literal values:
+
+    nodes { certname in ['foo', 'bar', 'baz'] }
+
+The operator can also be used to ensure the values of a field match the fields returned, from a subquery,
+which has the form of a nested PQL query within the filter:
+
+    nodes { certname in facts[certname] { value = 'foo' } }
+
+With the subquery form, we can even match on multiple fields, as long as the fields being matched, and the
+subqueries projection fields match:
+
+    facts { [name, value] in facts[name, value] { value ~ 'foo' } }
+
+#### Null detection: `is null`, `is not null`
+
+Null values in PuppetDB are treated differently to other values. So to detect if a field is a null, instead of
+doing an exact match comparison, you must use either the `is null` or `is not null` operator.
+
+To test if a field contains a `null`:
+
+    nodes { deactivated is null }
+
+Or conversely, to test if a field does not contain a `null` value:
+
+    nodes { deactivated is not null }
+
+#### Regexp array match: `~>`
+
+The array matches using the regular expressions provided within in each element. Array indexes are coerced
+to strings.
+
+For example. the following query would query the path element, matching any ethernet mac address:
+
+    fact_contents { path ~> ["networking", "eth.*", "macaddress"] }
+
+### Boolean operators
+
+Boolean operators are used within PQL filters to join conditons together to perform the filtering
+test within PuppetDB.
+
+There are only 3 boolean operators today, in order of natural precedence:
+
+* `!` - performs a logical `not` or negation
+* `and` - performs a logical `and` or conjuction
+* `or` - performs a logical `or` or disjunction
+
+### Grouping
+
+By default PQL binary operators are evaluated using the following natural order of precedence: `!`, `and`, and `or`.
+To override this precedence, you can group conditions together explicitly using parenetheses:
+
+    facts { name ~ '^operating' and ( name ~ 'system' or value = 'FlowerOS' ) }
+
+In this case the `or` gets evaluated before the `and` despite the natural order of precedence.
+
+You can nest as many levels of grouping as required.
+
+### Literal types
+
+Each field for an entity supports matching using a conditional against a provided literal value.
+
+#### Strings
+
+PQL supports legal UTF-8 strings as literals for comparison.
+
+There are two types of literal strings:
+
+* single quoted - no escaping, just straight text
+* double quoted - supports escape characters
```

Comment:
So ... a few reasons.

One of the things I notice about languages is that you often want the reverse of the quoting system you're using on the outside of a query string (like in curl) than you do on the inside. So think ... scripts ... bash ... that kind of thing. Having the ability to do that to avoid thinking too hard about escaping is sometime nice. In puppet for example, we support both, and sometimes users might wrap the string in single-quotes, but want a string inside the query, so they use double quotes:

    puppetdb::query("nodes { certname = 'foo' }")
    puppetdb::query('nodes { certname = "foo" }')

The other side of it, is for escaping versus non escaping behaviours. Obviously we should support escaping, as thats quite commonly required, things like carriage returns come to mind ... but sometimes you don't want escaping, single quotes will do.

I guess this feels like a normal/predictable things for me to do, but I can be convinced to drop one of them (perhaps the single quote) if you thought it was designing to far ahead.